observability/observability.go

```diff
@@ -0,0 +1,88 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package observability implements the tracing, metrics, and logging data
+// collection, and provides controlling knobs via a config file.
+//
+// Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+package observability
+
+import (
+	"context"
+
+	"google.golang.org/grpc/binarylog"
+	"google.golang.org/grpc/grpclog"
+	configpb "google.golang.org/grpc/observability/internal/config"
+)
+
+var (
+	logger = grpclog.Component("observability")
+	config *configpb.ObservabilityConfig
+)
+
+func init() {
+	internalInit()
+}
+
+// internalInit exists to allow unit tests to re-parse ENV var.
+func internalInit() {
+	config = parseObservabilityConfig()
+
+	// Logging is controlled by the config at methods level. Users might bring
+	// their in-house exporter. If logging is disabled, binary logging also
+	// won't start. The overhead should be minimum.
+	startLogging(config)
+}
+
+// Start is the opt-in API for gRPC Observability plugin. This function should
+// be invoked in the main function, and before creating any gRPC clients or
+// servers, otherwise, they might not be instrumented. At high-level, this
+// module does the following:
+//
+//   - it loads observability config from environment;
+//   - it registers default exporters if not disabled by the config;
+//   - it sets up binary logging sink against the logging exporter.
+//
+// Note: currently, the binarylog module only supports one sink, so using the
+// "observability" module will conflict with existing binarylog usage.
+// Note: handle the error
+func Start(ctx context.Context) error {
+	// If no project ID is found, that's ok
+	maybeUpdateProjectIDInObservabilityConfig(ctx, config)
+
+	// If the default logging exporter is not disabled, register one.
+	if config != nil && config.ExporterConfig.ProjectId != "" && !config.ExporterConfig.DisableDefaultLoggingExporter {
```

Comment:
As before.  Move happy path code to be outdented instead.

```go
if config.GetExporterConfig.GetProjectId() == "" || config.GetExporterConfig().GetDisableDefaultLoggingExporter() {
	return nil
}
....
```

You also avoid an explicit `nil` check this way.

Reply:
Done. Updated the code style to tackle the `nil` cases one-by-one.

The ProtoBuf getters... don't look like doing much work:

```
func (x *ObservabilityConfig) GetExporterConfig() *ObservabilityConfig_ExporterConfig {
	if x != nil {
		return x.ExporterConfig
	}
	return nil
}
```

I still included all the `nil` checks.

Reply:
The whole purpose of the `Get`ters is that first `if` block.  So you can write simpler code.  As explained above.